### ISSUE TITLE
[ENH] optimize pystr.upper() -> Py_UNICODE_TOUPPER(pystr)

### ISSUE THREAD

**Issue by jbrockmendel** [cython/cython#5215]
### Is your feature request related to a problem? Please describe.

In my code, I would like to do [...]
```cython
# add use case related code here
```

**Comment by 0dminnimda** [comment]
This is a deprecated function, I'd advise againts introducing it in the codebase

**Comment by scoder** [comment]
> This is a deprecated function, I'd advise against introducing it in the codebase

Agreed. Also, the `Py_UNICODE_TO*()` functions only apply to a single character. We already use these functions for single `Py_UCS4` character values, except for the three below:

https://docs.python.org/3/c-api/unicode.html#c.Py_UNICODE_TOUPPER
https://docs.python.org/3/c-api/unicode.html#c.Py_UNICODE_TOLOWER
https://docs.python.org/3/c-api/unicode.html#c.Py_UNICODE_TOTITLE

Recent CPython versions have implementation specific functions `_PyUnicode_ToUpperFull()`, `_PyUnicode_ToLowerFull()` and `_PyUnicode_ToTitleFull()`, which we could potentially start using, but I'm not sure it's worth it since the end result would still have to be a Python Unicode string object, given that we can receive up to four characters back.
